paddle/fluid/framework/details/build_strategy.cc

```diff
@@ -28,6 +33,15 @@ class ParallelExecutorPassBuilder : public ir::PassBuilder {
  public:
   explicit ParallelExecutorPassBuilder(const BuildStrategy &strategy)
       : ir::PassBuilder(), strategy_(strategy) {
+    // NOTE(dzh): memory optimize should be a runtime pass.
+    // However, after multi_devices_pass, VarHandle, OpHandle is
```

Comment:
VarHandle and OpHandle are just wrapper of ir Node. Every handle has a ir Node behind it.